elfi/methods/bo/acquisition.py

```diff
@@ -405,6 +407,95 @@ def evaluate_gradient(self, theta_new, t=None):
         return gradient
 
 
+class RandMaxVar(MaxVar):
+    r"""The randomised maximum variance acquisition method.
+
+    The next evaluation point is sampled from the variance of
+    the approximate posterior, i.e.:
+    \theta_{t+1} ~ V(p(\theta) * p_t(\theta | X_t)).
+
+    References
+    ----------
+    [1] arXiv:1704.00520 (Järvenpää et al., 2017)
+
+    """
+
+    def __init__(self, percentile_eps=None, *args, **opts):
+        """Initialise RandMaxVar.
+
+        Parameters
+        ----------
+        percentile_eps : int, optional
+            Percentile of the observed discrepancies used in setting the discrepancy threshold.
+
+        """
+        super(RandMaxVar, self).__init__(percentile_eps, *args, **opts)
+        self.name = 'rand_max_var'
+        self._n_nuts_samples = 100
+        self._limit_faulty_init = 10
+
+    def acquire(self, n, t=None):
+        """Acquire a batch of acquisition points.
+
+        Parameters
+        ----------
+        n : int
+            Number of acquisitions.
+        t : int, optional
+            Current iteration.
+
+        Returns
+        -------
+        array_like
+            Coordinates of the yielded acquisition points.
+
+        """
+        logger.debug('Acquiring the next batch of %d values', n)
+        gp = self.model
+
+        # Setting the discrepancy threshold.
+        if gp.Y is not None:
+            self.eps = np.percentile(gp.Y, self.percentile_eps)
+        else:
+            self.eps = 0.1
+
+        def _evaluate_gradient_logpdf(theta):
+            denominator = self.evaluate(theta)
+            if denominator == 0:
+                return -np.inf
+            pt_eval = self.evaluate_gradient(theta) / denominator
+            return pt_eval.ravel()
+
+        def _evaluate_logpdf(theta):
+            val_pdf = self.evaluate(theta)
+            if val_pdf == 0:
+                return -np.inf
+            return np.log(val_pdf)
+
+        # Obtaining the RandMaxVar acquisition.
+        for i in range(self._limit_faulty_init + 1):
+            if i > self._limit_faulty_init:
+                raise SystemExit("Unable to find a suitable initial point.")
+            # Proposing the initial point.
+            theta_init = np.zeros(shape=len(gp.bounds))
+            for idx_param, range_bound in enumerate(gp.bounds):
+                theta_init[idx_param] = self.random_state.uniform(range_bound[0], range_bound[1])
+            # Refusing to accept a faulty initial point.
+            if np.isinf(_evaluate_logpdf(theta_init)):
+                continue
+            # Sampling using NUTS.
+            samples = mcmc.nuts(self._n_nuts_samples,
+                                theta_init,
+                                _evaluate_logpdf,
+                                _evaluate_gradient_logpdf,
+                                seed=self.seed)
+            # Using the last n points of the NUTS chain for the acquisition batch.
+            batch_theta = samples[-n:, :]
```

Comment:
Should ensure that `self._n_nuts_samples` >= `n`, maybe already when initializing?

Reply:
Good catch, fixed